Rework backend list

fd0 opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

### What is the purpose of this change? What does it change?

It changes the signature of the `List()` functions of the `Backend` and `Repository` interfaces to take a function (similar to `filepath.Walk()`) and return not only a name/ID, but also a size.

### Was the change discussed in an issue or in the forum before?

https://github.com/restic/restic/issues/1567


### TODO

 * [x] change the `List()` function for `Repository` and `Backend`
 * [x] fix all uses of `List()`
 * [x] change all backends to provide size information
 * [x] amend the REST protocol to provide size information directly in the list response
 * [x] implement the REST v2 protocol in the backend

fd0 committed:
backend: Rework List()

For a discussion see #1567

fd0 committed:
backend: Add tests for new List() function

fd0 committed:
Change List() implementation for all backends

fd0 committed:
Fix calls to repo/backend.List() everywhere

fd0 committed:
Vendor golang.org/x/sync

zcalusic commented:
@fd0, I can't find where is the needed change in REST API documented in these commits. Maybe I'm just lost. :(

REST API documentation is my primary guide to what needs to be implemented in rest-server. ;)

fd0 commented:
Sorry, I haven't had the time to propose something, so it's not yet in the code here

fd0 committed:
Document the REST v2 protocol

ifedorenko commented on internal/repository/repository.go:
```diff
@@ -536,22 +536,15 @@ func (r *Repository) KeyName() string {
-				}
-			}
+// List runs fn for all files of type t in the repo.
+func (r *Repository) List(ctx context.Context, t restic.FileType, fn func(restic.ID, int64) error) error {
```
Why callback function gets two separate `restic.ID` and `int64` parameters, and not single `restic.FileInfo` parameter?

fd0 replied:
`restic.FileInfo` contains `Name string`, which is fine for the backend: There, the files have names. On the "repository" abstraction level, we're only dealing with `restic.ID`s. I did not see the point in combining the two parameters into a new struct, as I did not come up with a good name for it.

Do you have an idea?

Thinking about it, it may get easier to extend should we ever decide to add a third parameter. hm.

fd0 replied:
We could also have both, `backend.FileInfo` and `repository.FileInfo`, but that may be too confusing...

ifedorenko replied:
Ah, didn't notice `restic.ID` vs `string` difference. In that case, I think two parameters are a good decision (no sense to worry about hypothetical third parameter until we actually need it, IMO).

fd0 replied:
Agreed

ifedorenko commented on internal/restic/backend.go:
```diff
@@ -32,10 +32,9 @@ type Backend interface {
-	List(ctx context.Context, t FileType) <-chan string
+	// List runs fn for each file in the backend which has the type t. When an
+	// error occurs (or fn returns an error), List stops and returns it.
+	List(ctx context.Context, t FileType, fn func(FileInfo) error) error
```
I would mention that `fn` is allowed to call Backend. This has implications for how backend implementations throttle concurrent remote calls. 

I would also explicitly say that `fn` will be called exactly once for each `FileInfo` during successful `List()` execution (and at most once in case of an error). This has implications for `BackendRetry` implementation.

ifedorenko replied:
Hmm, I think we also need to explicitly state that `fn` will be called on the calling gorouting. Otherwise each client will need to deal with synchronization issues if they want to [collect results in a map](https://stackoverflow.com/questions/11063473/map-with-concurrent-access) for example.

fd0 replied:
Ah, good points, thanks. The `List()` backend call isn't restricted (in terms of parallel connections) in any existing backend, only the other operations like `Load()` and `Save()`.

> Hmm, I think we also need to explicitly state that fn will be called on the calling gorouting

Why do you feel that way? Sure, we can add that, but In my experience, APIs for libraries rather state if something isn't run in the calling goroutine...

ifedorenko replied:
If "on the same gorouting" is the normal/expected way of doing things in Go, then indeed no need to mention it explicitly. (sorry, I am new to Go, still learning my ways around).

fd0 replied:
We can still add it for clarity, I was just wondering if maybe I got something wrong :)

ifedorenko reviewed (commented):
Looks  good, with couple of minor comments/questions. I'll try to update BackendRetry and my OneDrive backend to use the new API in the next few days.

fd0 commented on internal/backend/test/tests.go:
```diff
@@ -312,6 +330,122 @@ func (s *Suite) TestList(t *testing.T) {
+			return nil
+		})
+
+		if errors.Cause(err) != context.Canceled {
```
Ok, that's something we need to add to our `errors` package, here: https://github.com/restic/restic/blob/a3d43a92b327e9aba596ac4d4d42f6ed54989255/internal/errors/wrap.go#L6

I'll do that later.

fd0 commented on internal/backend/test/tests.go:
```diff
@@ -312,6 +330,122 @@ func (s *Suite) TestList(t *testing.T) {
+		ctx, cancel := context.WithCancel(context.TODO())
+		defer cancel()
+
+		timeout := 200 * time.Millisecond
```
Yeah, we need to find a proper solution here and not depend on timing. This test case has bitten me once already (failed on Travis on darwin), so let's do this properly.

ifedorenko replied:
We have to assume all remote backend tests will fail regularly. We just need to wrap each test in retry-with-backoff loop and hope the failures are not intermittent bugs in restic. For this specific test I'd increase timeout to 500ms and make the timeout per-backend configurable.

fd0 replied:
Ah, I was thinking about a more robust approach, like this:

```go
        // rather large timeout
        timeout := time.Second

        ctxTimeout, _ := context.WithTimeout(ctx, timeout)

        i := 0
 		// pass in a context with a small timeout
 		err := b.List(ctxTimeout, restic.DataFile, func(fi restic.FileInfo) error {
 			i++
  
 			// wait until the context is timed out, so the function is not called again
            <-ctxTimeout.Done()
 			return nil
 		})
```

Maybe we can then even accept 0 or 1 calls of the function (so the timeout happens either before the first call of the function, or before the second call).

Thoughts?

ifedorenko replied:
> accept 0 or 1 calls of the function

yes, good idea, haven't thought about it.

fd0 committed:
Add comments for List()

fd0 committed:
errors: Make Cause() unwrap *url.Error

fd0 committed:
backend/test: Reliably trigger timeout error

fd0 committed:
repository: Fix repack test

This reduces the chance of duplicate blobs, otherwise the tests fail
(make the contents of a blob depend on a pseudo-random number instead of
the size, sizes may be duplicate).

fd0 commented:
I've added the changes we discussed, including a clarification for the REST protocol.

fd0 committed:
backend/rest: Implement REST API v2

fd0 commented:
We've had a discussion on the API protocol version selection for the REST API, and settled on using an HTTP request header (`Accept`) and a proper mime type.

This PR is done now, any objections?
